File: src/core/crdt.ts
/** RGA: descending Lamport sibling order; a tombstone never hides descendants. */
export type Op =
  | { id: string; kind: "insert"; after: string; value: string }
  | { id: string; kind: "delete"; target: string };
export const ROOT = "";
const pattern = /^([1-9][0-9]{0,14}):([a-zA-Z0-9_-]{1,64})$/;
export function clock(id: string): number {
  const m = pattern.exec(id);
  if (!m) throw new Error("Invalid operation ID");
  return Number(m[1]);
}
export function compare(a: string, b: string): number {
  const n = clock(a) - clock(b);
  return n || (a < b ? -1 : a > b ? 1 : 0);
}
export function validate(op: Op): void {
  if (!op || typeof op.id !== "string") throw new Error("Invalid operation");
  clock(op.id);
  if (op.kind === "insert") {
    if (
      typeof op.value !== "string" ||
      [...op.value].length !== 1 ||
      /[\uD800-\uDFFF]/u.test(op.value)
    )
      throw new Error("Insert must contain one Unicode scalar");
    if (
      typeof op.after !== "string" ||
      (op.after !== ROOT && clock(op.after) >= clock(op.id))
    )
      throw new Error("Parent must precede insertion");
  } else if (op.kind === "delete") {
    if (typeof op.target !== "string" || clock(op.target) >= clock(op.id))
      throw new Error("Target must precede deletion");
  } else throw new Error("Unknown operation");
}
export function equal(a: Op, b: Op): boolean {
  return (
    a.kind === b.kind &&
    a.id === b.id &&
    (a.kind === "insert" && b.kind === "insert"
      ? a.after === b.after && a.value === b.value
      : a.kind === "delete" && b.kind === "delete" && a.target === b.target)
  );
}
export class CRDT {
  private ops = new Map<string, Op>();
  private time = 0;
  constructor(readonly actor: string) {
    if (!/^[a-zA-Z0-9_-]{1,64}$/.test(actor)) throw new Error("Invalid actor");
  }
  apply(batch: Op[]): void {
    const staged = new Map<string, Op>();
    for (const op of batch) {
      validate(op);
      const old = staged.get(op.id) ?? this.ops.get(op.id);
      if (old && !equal(old, op)) throw new Error("Conflicting operation ID");
      staged.set(op.id, op);
    }
    for (const [id, op] of staged) {
      this.ops.set(id, { ...op });
      this.time = Math.max(this.time, clock(id));
    }
  }
  all(): Op[] {
    return [...this.ops.values()].map((op) => ({ ...op }));
  }
  visible(): { id: string; value: string }[] {
    const children = new Map<string, Extract<Op, { kind: "insert" }>[]>();
    const deleted = new Set<string>();
    for (const op of this.ops.values())
      if (op.kind === "delete") deleted.add(op.target);
      else {
        const list = children.get(op.after) ?? [];
        list.push(op);
        children.set(op.after, list);
      }
    for (const list of children.values())
      list.sort((a, b) => compare(b.id, a.id));
    const stack = [...(children.get(ROOT) ?? [])].reverse();
    const out: { id: string; value: string }[] = [];
    while (stack.length) {
      const n = stack.pop()!;
      if (!deleted.has(n.id)) out.push({ id: n.id, value: n.value });
      const list = children.get(n.id) ?? [];
      for (let i = list.length - 1; i >= 0; i--) stack.push(list[i]);
    }
    return out;
  }
  text(): string {
    return this.visible()
      .map((n) => n.value)
      .join("");
  }
  edit(next: string): Op[] {
    const old = this.visible(),
      chars = [...next];
    let start = 0,
      end = 0;
    while (
      start < old.length &&
      start < chars.length &&
      old[start].value === chars[start]
    )
      start++;
    while (
      end < old.length - start &&
      end < chars.length - start &&
      old[old.length - 1 - end].value === chars[chars.length - 1 - end]
    )
      end++;
    const batch: Op[] = [];
    const id = () => `${++this.time}:${this.actor}`;
    for (const n of old.slice(start, old.length - end))
      batch.push({ kind: "delete", id: id(), target: n.id });
    let after = start ? old[start - 1].id : ROOT;
    for (const value of chars.slice(start, chars.length - end)) {
      const op: Op = { kind: "insert", id: id(), after, value };
      batch.push(op);
      after = op.id;
    }
    this.apply(batch);
    return batch;
  }
}

File: internal/crdt/crdt.go
// Package crdt is an independent RGA reference implementation.
package crdt

import (
	"errors"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"unicode/utf8"
)

type Op struct {
	ID     string `json:"id"`
	Kind   string `json:"kind"`
	After  string `json:"after"`
	Value  string `json:"value,omitempty"`
	Target string `json:"target,omitempty"`
}

var idPattern = regexp.MustCompile(`^[1-9][0-9]{0,14}:[a-zA-Z0-9_-]{1,64}$`)

func Clock(id string) (int64, error) {
	if !idPattern.MatchString(id) {
		return 0, errors.New("invalid operation ID")
	}
	n, _ := strconv.ParseInt(strings.SplitN(id, ":", 2)[0], 10, 64)
	return n, nil
}
func Validate(op Op) error {
	n, err := Clock(op.ID)
	if err != nil {
		return err
	}
	switch op.Kind {
	case "insert":
		if !utf8.ValidString(op.Value) || utf8.RuneCountInString(op.Value) != 1 || op.Target != "" {
			return errors.New("invalid insertion")
		}
		if op.After != "" {
			p, e := Clock(op.After)
			if e != nil || p >= n {
				return errors.New("invalid parent")
			}
		}
	case "delete":
		p, e := Clock(op.Target)
		if e != nil || p >= n || op.Value != "" || op.After != "" {
			return errors.New("invalid deletion")
		}
	default:
		return errors.New("unknown operation")
	}
	return nil
}

type Document struct{ operations map[string]Op }

func New() *Document { return &Document{operations: map[string]Op{}} }
func (d *Document) Apply(ops []Op) error {
	staged := map[string]Op{}
	for _, op := range ops {
		if err := Validate(op); err != nil {
			return err
		}
		if old, ok := d.operations[op.ID]; ok && old != op {
			return errors.New("conflicting operation ID")
		}
		if old, ok := staged[op.ID]; ok && old != op {
			return errors.New("conflicting operation ID")
		}
		staged[op.ID] = op
	}
	for id, op := range staged {
		d.operations[id] = op
	}
	return nil
}
func (d *Document) Text() string {
	children := map[string][]Op{}
	deleted := map[string]bool{}
	for _, op := range d.operations {
		if op.Kind == "delete" {
			deleted[op.Target] = true
		} else {
			children[op.After] = append(children[op.After], op)
		}
	}
	for key := range children {
		sort.Slice(children[key], func(i, j int) bool {
			a, b := children[key][i].ID, children[key][j].ID
			an, _ := Clock(a)
			bn, _ := Clock(b)
			if an == bn {
				return a > b
			}
			return an > bn
		})
	}
	stack := []Op{}
	push := func(list []Op) {
		for i := len(list) - 1; i >= 0; i-- {
			stack = append(stack, list[i])
		}
	}
	push(children[""])
	var out strings.Builder
	for len(stack) > 0 {
		n := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if !deleted[n.ID] {
			out.WriteString(n.Value)
		}
		push(children[n.ID])
	}
	return out.String()
}
